make/mt-os/mt-esp32.mk

```diff
@@ -5,13 +5,76 @@
 
 include make/mt-os/mt-os-common.mk
 
-#$(IDF_PATH) This is exported in the shell as as IDF requirement
-XI_ESP_IDF_SDK_PATH ?= $(IDF_PATH)
-
 # The path to CC and AR must be in the environment $PATH as an IDF requirement
 CC = xtensa-esp32-elf-gcc
 AR = xtensa-esp32-elf-ar
 
+XI_ESP32_PREREQUISITE_DOWNLOAD_PATH := $(HOME)/Downloads/esp32
+
+################################
+# auto-provide ESP32 SDK #######
+################################
+ifeq (,$(wildcard $(IDF_PATH)))
+    $(info ESP32 SDK is NOT available, using auto-downloaded ESP32 SDK from $(XI_ESP32_PREREQUISITE_DOWNLOAD_PATH))
+
+    XI_ESP_IDF_SDK_PATH := $(XI_ESP32_PREREQUISITE_DOWNLOAD_PATH)/esp-idf
+
+$(XI_ESP_IDF_SDK_PATH):
+	@-mkdir -p $@
+	git clone --recursive -b release/v2.1 https://github.com/espressif/esp-idf.git $(XI_ESP_IDF_SDK_PATH)
+
+XI_BUILD_PRECONDITIONS += $(XI_ESP_IDF_SDK_PATH)
+
+else
+
+#$(IDF_PATH) This is exported in the shell as as IDF requirement
+XI_ESP_IDF_SDK_PATH ?= $(IDF_PATH)
+
+endif
+
+################################
+# auto-provide ESP32 toolchain #
+################################
+XI_ESP32_AVAILABILITY_CHECK_CC := $(shell which $(CC) 2> /dev/null)
+
+ifndef XI_ESP32_AVAILABILITY_CHECK_CC
+    $(info ESP32 compiler is NOT available, using auto-downloaded ESP32 toolchain from $(XI_ESP32_PREREQUISITE_DOWNLOAD_PATH))
+
+    CC := $(XI_ESP32_PREREQUISITE_DOWNLOAD_PATH)/xtensa-esp32-elf/bin/$(CC)
+    AR := $(XI_ESP32_PREREQUISITE_DOWNLOAD_PATH)/xtensa-esp32-elf/bin/$(AR)
+
+ifeq ($(XI_HOST_PLATFORM),Darwin)
+    # osx cross-compilation toolchain downloads
+
+    XI_ESP32_TOOLCHAIN_DOWNLOAD_FILE := ~/Downloads/esp32/xtensa-esp32-elf-osx-1.22.0-73-ge28a011-5.2.0.tar.gz
+
+    XI_ESP32_TOOLCHAIN_URL := https://dl.espressif.com/dl/xtensa-esp32-elf-osx-1.22.0-73-ge28a011-5.2.0.tar.gz
+
+else ifeq ($(XI_HOST_PLATFORM),Linux)
+
+    XI_ESP32_TOOLCHAIN_DOWNLOAD_FILE := ~/Downloads/esp32/xtensa-esp32-elf-linux64-1.22.0-73-ge28a011-5.2.0.tar.gz
+
+    XI_ESP32_TOOLCHAIN_URL := https://dl.espressif.com/dl/xtensa-esp32-elf-linux64-1.22.0-73-ge28a011-5.2.0.tar.gz
+
+endif
+
+XI_BUILD_PRECONDITIONS += $(CC)
+
+$(XI_ESP32_TOOLCHAIN_DOWNLOAD_FILE):
+	@echo XI ESP32 BUILD: downloading xtensa esp32 toolchain to file $@
+	@-mkdir -p $(dir $@)
+	@curl -L -o $@ $(XI_ESP32_TOOLCHAIN_URL)
+
+$(CC): $(XI_ESP32_TOOLCHAIN_DOWNLOAD_FILE)
+	@echo XI ESP32 BUILD: extracting xtensa esp32 toolchain to have compiler $(CC)
+	@tar -xf $< -C ~/Downloads/esp32
+	touch $@
+
+# XI_ESP32_AVAILABILITY_CHECK_CC
+endif
+
+include make/mt-config/mt-tls-wolfssl-esp32.mk
```

Comment:
The IDF SDK includes a port of mbedTLS with hardware acceleration.
I moved to wolfSSL before being able to fix our BSP for that version of mbedTLS, but I'm sure some users will be interested in taking advantage of that acceleration.

Maybe we should `ifeq ($(XI_BSP_TLS),wolfssl)` the inclusion of `mt-tls-wolfssl-esp32.mk` so it's easier for users to move towards mbedTLS. Otherwise, finding this line may be tricky for them. What do you think?

Reply:
this is good point, trying to figure out the solution. Maybe it should depend on PRESET, thinking...

Reply:
I think this will do the job:

```
ifeq ($(XI_BSP_TLS),wolfssl)
include make/mt-config/mt-tls-wolfssl-esp32.mk
endif
```

I just ran it in your branch (modified `wolfssl` with `mbedtls` in the ifeq) and it worked as expected.

Don't let this hold up the release, though. If you're not sure this is the best approach, feel free to merge and we can figure this out later.